File: python/gravel/src/artanis_gravel/manifest/hash.py
"""Prompt content hash normalization.

Identical algorithm to packages/sdk-ts/src/manifest/hash.ts. CI verifies
output equivalence.
"""
from __future__ import annotations

import hashlib
import re
import secrets


def normalize(text: str) -> str:
    # 1. Convert line endings to \n.
    out = text.replace("\r\n", "\n").replace("\r", "\n")
    # 2. Strip trailing whitespace on each line.
    out = "\n".join(re.sub(r"[ \t]+$", "", line) for line in out.split("\n"))
    # 3. Strip leading + trailing blank lines.
    out = re.sub(r"^(\s*\n)+", "", out)
    out = re.sub(r"(\n\s*)+$", "", out)
    return out


def hash_prompt(text: str) -> str:
    return "sha256:" + hashlib.sha256(normalize(text).encode("utf-8")).hexdigest()


def generate_prompt_id(path: str, char_start: int | None = None) -> str:
    seed = f"{path}:{char_start if char_start is not None else 'file'}:{secrets.token_hex(8)}"
    return "p_" + hashlib.sha1(seed.encode("utf-8")).hexdigest()[:12]

File: packages/sdk-ts/src/manifest/hash.ts
/**
 * Hash normalization for prompt content. Whitespace differences shouldn't
 * change a hash. Spec: gravel-cloud/docs/spec/manifest.md §4.
 *
 * Identical to python/gravel/src/artanis_gravel/manifest/hash.py — kept in sync.
 */
import { createHash } from 'node:crypto'

export function normalize(text: string): string {
  // 1. Convert line endings to \n.
  let normalized = text.replace(/\r\n?/g, '\n')
  // 2. Strip trailing whitespace on each line.
  normalized = normalized
    .split('\n')
    .map((line) => line.replace(/[ \t]+$/, ''))
    .join('\n')
  // 3. Strip leading + trailing blank lines.
  normalized = normalized.replace(/^(\s*\n)+/, '').replace(/(\n\s*)+$/, '')
  return normalized
}

export function hashPrompt(text: string): string {
  return 'sha256:' + createHash('sha256').update(normalize(text)).digest('hex')
}

/**
 * Stable ID for a prompt entry. Generated once at first detection and persists
 * thereafter (path-rename-safe).
 */
export function generatePromptId(path: string, charStart?: number): string {
  const seed = `${path}:${charStart ?? 'file'}:${process.hrtime.bigint()}:${Math.random()}`
  // 8-char prefix is plenty for collision avoidance within one repo.
  return 'p_' + createHash('sha1').update(seed).digest('hex').slice(0, 12)
}
